src/shell/shell_cmds.cpp

```diff
@@ -2572,3 +2573,121 @@ void DOS_Shell::CMD_MOVE(char* args)
 		}
 	}
 }
+
+static std::vector<std::string> search_files(std::string_view query)
+{
+	const auto save_dta = dos.dta();
+	dos.dta(dos.tables.tempdta);
+	const auto dta = DOS_DTA(dos.dta());
+
+	bool found = DOS_FindFirst(std::string(query).c_str(),
+	                           ~(DOS_ATTR_VOLUME | DOS_ATTR_DIRECTORY));
+
+	std::vector<std::string> files = {};
+	while (found) {
+		DOS_DTA::Result result = {};
+		dta.GetResult(result);
+		files.emplace_back(std::move(result.name));
+		found = DOS_FindNext();
+	}
+
+	dos.dta(save_dta);
+	
+	return files;
+}
+
+void DOS_Shell::CMD_FOR(char* args)
+{
+	HELP("FOR");
+	
+	static const auto delimiters = std::string_view(",;= \t");
```

Comment:
Just use `std::string` or`const char *` everywhere, please. This is not performance intensive code, let's do the simplest thing.

Reply:
Yup, normally the char* string literals are fine as:

`constexpr auto str = "....";`, but in this case you want the class benefits of std::string so it can upgraded:

```c++
constexpr std::string_view delimiters = "....";
```

The nice thing with strong literals is that they're permanently pinned I to the program, so this is a very simple and elegant way to get the stringview benefit without any risks.

Reply:
Can you add a top level comment describing the  syntax that this function digests - including a real life example string that hits all the functional bits in the body?

Then include more comments below describing which chunk of the syntax each bit of logic is digesting, one by one, to make it easier to follow along. 

(Right now, I'm trying to reverse-engineer it in my head from the delimiter and search strings, and getting too overwhelmed to hold it all. My contest space is like ChatGPT 0.5 beta level 😅)

Reply:
My other major comment is that the entire body is using totally generic variable names, often simply named after the type itself. 

- args, variable, token, string_list, str, list_start, list_end, and parameters. Can these be named for the chunk or bits of the FOR loop components themselves? 

Or maybe they are named after the FOR components, but I just need help connecting the dots.

Reply:
> How exactly is string_view not the simplest thing here?

Well, the simplest thing for me is this:

```ini
static const char* blah delimiters = ",;= \t";
```

or (if you need `std::string` methods):

```ini
static const std::string delimiters = ",;= \t";
```

Seeing `string_view` anywhere in the code kinda signals to me: "ok, this is performance intensive code, we're probably dealing with large strings, maybe many thousands of string manipulations in tight loops, so we want to avoid copies because it makes a big difference, etc." None of that happening here? Also it signals me to be careful not to let the `string_view` "leak out" from the function/module/etc because a `string_view` is basically a pointer-pair.

So I wouldn't call it the simplest thing to do 😄

Reply:
> `static const std::string delimiters = ",;= \t";`

This cannot be a compile time constant however, unlike `string_view`.

Reply:
> Seeing string_view anywhere in the code kinda signals to me: "ok, this is performance intensive code".

But it shouldn't; well, the purpose of using it isn't solely for higher performance code. It's just a functional wrapper to strap on top of raw C-string. 

It says to me, "I only intent to perform read only operations to find / compare / check" using nicer C++ string calls instead of more dangerous.

It tells me were inspecting as opposed to authoring new strings.

The RHS is a const char[], so string_view is a perfect fit to make it a read

Reply:
(You can imagine a `std::vector_view<T>`, that takes in a C-array and lets you run .size(), .find(Val), .any_of(Val), etc..)

Reply:
> > Seeing string_view anywhere in the code kinda signals to me: "ok, this is performance intensive code".
> 
> It's just a functional wrapper to strap on top of a read-only C-string.
> 
> It says to me, _"I only intent to perform read only operations to find / compare / check"_ using nicer C++ string calls instead of more dangerous C calls.
> 
> Basically: _inspecting_ as opposed to _authoring_ new strings.
> 
> We can inspect with std::string, too: but string_view makes that use-case guaranteed and clearer.

Yeah, and actually using string views as function args is one of the nicer use cases where it acts as a sort of a "wildcard" type that can accept both `std::string` and `const char *`.

I guess if you only use them internally and never store them in class members, etc., or only where the lifecycles of the actual strings it references are very clear, then they're ok, and no doubt can be helpful.

My problem with them is that once you use string views, you also need to consider the lifecycle of the underlying strings, and that's actually a source for bugs and/or crashes as we've seen. So to be 100% clear, _that_ is my only (and biggest at the same time) problem with them that it kinda forces us back into the "manual memory management" mindset. Unless I'm missing something obvious 😄 With references or straight copies that's never a problem.